Add Assistant chat interface

derekblank opened the pull request:
Resolves Automattic/dotcom-forge#7323

## Proposed Changes

- Updates ContentTabAssistant to display chat messages
- Adds `useAssistant` hook to store and retrieve chat messages (currently using localStorage, but expected to change)
- Introduces basic responsive design functionality 
- Updates general spacing of all tabs
- (For testing) Simulates a server response
- Moves icons to an icons directory


https://github.com/Automattic/studio/assets/643285/348f240f-92cd-4768-a8ac-e9ada1bdeecd



## Testing Instructions
Proposed requirements, from https://github.com/Automattic/dotcom-forge/issues/7323:

**When I click on AI Assistant tab, I can see the chat interface**
1. Start Studio and add site
2. Navigate to the Assistant tab and observe a text input with an empty chat window

**Pressing enter on the text input will send the message**
1. Start Studio and add site
2. Navigate to the Assistant tab and type a message
3. Hit enter and observe the message appears in the chat window
4. Observe a simulated server response

**The messages are stored locally**
1. In the left Sidebar, select a different site and navigate to that site's Assistant tab
2. Observe that the chat is empty (for a new site)
4. Start a new chat in a different site, or close the app and reopen it
5. Observe the chat is persisted (alternatively, use the Developer Tools to view the localStorage contents)

**When I change the selected site, I can see a new chat interface with the messages I sent related to that site**
1. In the left Sidebar, select a different site and navigate to that site's Assistant tab
2. Observe that the chat is empty (for a new site)
4. Start a new chat in a different site, or close the app and reopen it
5. Observe the chat is persisted (alternatively, use the Developer Tools to view the localStorage contents)

**The clear chat button will delete the messages for that site**
1. Create a chat conversation with several messages
2. Click the three-dots icon to the right of the text input
3. Observe that the chat is cleared from the window


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Have you checked for TypeScript, React or other console errors?

## Todo
- [x] Check for any layout style regressions, including Windows and RTL layouts
- [ ] ~Add the user id as a key to the chat storage so that chats are unique to a user's account~ _Deferred to_ https://github.com/Automattic/dotcom-forge/issues/7532
- [x] Seek and incorporate design feedback on the current UX

derekblank committed:
Update AI Assistant tab to handle text input and chat interface

derekblank committed:
Update AI Assistant chat interface tests

derekblank committed:
Update Assistant chat bubble styles

derekblank committed:
Apply padding to individual content tabs to preserve spacing

derekblank committed:
Update Assistant tab scrolling

derekblank committed:
Add useAssistant hook and tests

derekblank committed:
Fix clear conversation behavior

derekblank committed:
Update Icon file linting

derekblank commented:
@sejas Opening up the PR to get feedback on the initial approach. Please note when testing that server responses are simulated, and that the UX is iterative and expected to change as more elements are introduced. This PR focused on the five requirements listed in https://github.com/Automattic/dotcom-forge/issues/7323 with any additional features expected to follow, but open to any other changes we'd want to include in this initial pass.

derekblank commented:
@matt-west Any feedback on the design and UX is very welcome. The current design UX in this PR is a general composite of the different [Figma iterations](https://www.figma.com/design/RToz6tIuQ7nlZrikBte4GU/Studio?node-id=3074-116532&t=8OlDwmvRo4U4MwD4-0), and could be expected to evolve as we expand the responsive UI (and communication with the backend).

If you're able to run this branch locally or test with one of the test builds, I'd be particularly interested in your thoughts on the current scrolling UX. I've updated the styling so that the Assistant tab can scroll on the vertical axis, while other tabs should remain static. On the Assistant tab, the text input area remains fixed to the bottom of the window, while the header and tab navigation scroll with the chat. It's possible to make the site header and/or the navigation tabs fixed to the top of the window when the Assistant tab scrolls as well. Also interested in general spacing and layout updates we'd want to make on this first pass, or any other details to include.

derekblank commented on src/components/content-tab-assistant.tsx:
```diff
@@ -1,44 +1,115 @@
+			<div
+				data-testid="assistant-input"
+				className="px-8 py-6 bg-white flex items-center border-t border-gray-200"
+				style={ { position: 'sticky', bottom: 0 } }
```
If we keep the input area fixed to the bottom of the screen in the UX, these styles can be moved to somewhere more permanent or reusable.

dcalhoun replied:
Would it make sense to utilize TailwindCSS' `sticky bottom-0` for consistency?

dcalhoun replied:
Would it make sense to utilize TailwindCSS' `sticky bottom-0` for consistency?

derekblank commented on src/components/content-tab-assistant.tsx:
```diff
@@ -1,44 +1,115 @@
+	const [ input, setInput ] = useState< string >( '' );
+	const endOfMessagesRef = useRef< HTMLDivElement >( null );
+
+	const simulatedResponses = [
+		'Welcome to the Studio assistant. I can help manage your site, debug issues, and navigate your way around the WordPress ecosystem.',
+		'What can I help you with today?',
+		"To install the Jetpack plugin in WordPress, you can follow these steps: \n\n1. Log in to your WordPress admin dashboard. \n2. Go to 'Plugins' > 'Add New'. \n3. In the search bar, type 'Jetpack' and press Enter. \n4. Find the Jetpack plugin in the search results (it should be the first one) and click 'Install Now'. \n5. After the plugin is installed, click 'Activate'.",
+		'If you prefer to install it via WP-CLI, you can do so with the following command: wp plugin install jetpack --activate',
+		"After installing and activating Jetpack, you'll need to connect it to a WordPress.com account. \nThis is necessary because Jetpack uses the WordPress.com infrastructure to provide many of its features. You can do this from the Jetpack settings page in your WordPress admin dashboard.",
+	];
```
Server responses are simulated until conversations from https://github.com/Automattic/dotcom-forge/issues/7427 can be incorporated.

sejas replied:
We can integrate the API calls directly instead! I'll create a quick PR.

derekblank replied:
> We can integrate the API calls directly instead!

Perfect, even better. 👍

derekblank commented on src/components/content-tab-assistant.tsx:
```diff
@@ -1,44 +1,115 @@
+		<div className="inline-block p-2 rounded-sm border border-gray-300 bg-white lg:max-w-[70%] user-select-text">
+			{ text }
+		</div>
+	</div>
```
Potentially, both `UserMessage` and `ResponseMessage` could be a single component, or live in different files from the ContentTabAssistant. I debated grouping everything into its own dedicated `assistant` subdirectory, but felt it was not needed quite yet. Open to feedback on code organization here.

sejas replied:
Yeah, I think having only one component makes sense, and let's use `{children}` instead of just the text.

Here is a suggested code:

```
interface MessageProps {
	children: ReactNode;
	isUser: boolean;
}

export const Message = ( { children, isUser }: MessageProps ) => (
	<div className={ cx( 'flex mb-2 mt-2', isUser ? 'justify-end' : 'justify-start' ) }>
		<div
			className={ cx(
				'inline-block p-2 rounded-sm border border-gray-300 lg:max-w-[70%] user-select-text',
				! isUser && 'bg-white'
			) }
		>
			{ children }
		</div>
	</div>
);
```

derekblank replied:
Updated, thanks for the suggestion.

derekblank commented on src/components/app.tsx:
```diff
@@ -41,7 +41,7 @@ export default function App() {
 				<main
 					data-testid="site-content"
-					className="py-8 pr-8 bg-white overflow-y-auto h-full flex-grow rounded-chrome"
+					className="bg-white h-full flex-grow rounded-chrome overflow-hidden"
```
These padding updates to the other tabs are to accommodate the Assistant tab having a different background color than the other three tabs. The general main window padding has been moved off the `<main />` window element to the individual tabs themselves.

sejas replied:
I think we can remove this `overflow-hidden`

derekblank commented on src/components/content-tab-assistant.tsx:
```diff
@@ -1,44 +1,115 @@
-				<button className="btn btn-primary">Send</button>
+				<button aria-label="menu" className="p-2 ml-2 cursor-pointer" onClick={ clearInput }>
+					<MenuIcon />
+				</button>
```
This button clears the chat for demonstration purposes, but is expected to expand into a menu component that contains other menu options, as per the designs.

dcalhoun replied:
When we pursue final designs, it might be beneficial to use the WordPress `Button` and `Input` components, intead of `button` and `input` respectively, to receive the focus styles we leverage elsewhere in the UI.

derekblank replied:
Good idea. I went ahead and updated the Assistant tab to use Button in this PR.

Per the intended designs, I noted the Assistant Icon and menu button are included within the input "container". I also noted that the equivalent WordPress input component, [TextControl](https://developer.wordpress.org/block-editor/reference-guides/components/text-control/), appears to not support passing children, so I think it's best to make that change in a follow-up PR.

derekblank commented on src/hooks/use-assistant.ts:
```diff
@@ -0,0 +1,34 @@
+	const [ messages, setMessages ] = useState< Message[] >( [] );
+
+	useEffect( () => {
+		const storedChat = localStorage.getItem( selectedSiteId );
```
The user id could be added here as an extra info to the data key to make chats unique per user. Open to feedback on the approach of adding a user id value to the key here, or in a separate PR.

kozer reviewed (approved):
Tested it and work as expected. More specifically:
- Send a message and received a response from the simulated responses
- Change site, and noticed that chat is empty. Send message in the other site, and also worked.
- Chat messages persisted, even after a restart
- Clicking the "3 dots" cleared the chat.

Nice work @derekblank !

sejas commented on src/index.css:
```diff
@@ -84,3 +84,8 @@ blockquote {
+/* Allow user selected text on specified elements */
+.user-select-text {
+	user-select: text;
+}
```
Should we use `select-text` or even `select-all` from Tailwind?
https://tailwindcss.com/docs/user-select#allowing-text-selection

dcalhoun commented on src/components/content-tab-assistant.tsx:
```diff
@@ -1,44 +1,115 @@
+export const UserMessage = ( { text }: { text: string } ) => (
+	<div className="flex justify-end mb-2 mt-2">
+		<div className="inline-block p-2 rounded-sm border border-gray-300 lg:max-w-[70%] user-select-text">
+			{ text }
```
To improve the comprehensibility of this interface for users relying upon assitive technologies, we might consider labeling each message with the author. E.g., `aria-label="Your message, Install a theme"`, `aria-lable="Studio AI, You can install..."`. 

Additionally, we might consider announcing new messages as they arrive using [`@wordpress/a11y`](https://github.com/WordPress/gutenberg/tree/18ca36c0102b1384f0392e8d9d9e3db9c0f7cc02/packages/a11y#accessibility-a11y), which we [already leverage](https://github.com/search?q=repo%3AAutomattic%2Fstudio%20%40wordpress%2Fa11y&type=code) in the Studio project.

sejas reviewed (commented):
Great job! I tested regressions and couldn't find any.

I left a few optional comments, and it would be great if we can make the messages to scroll instead of the whole window.

If fixing the scroll is hard, then we can merge this PR and create the fix in a different PR.

## Current scroll


https://github.com/Automattic/studio/assets/779993/4f0b60f0-542b-4327-91c3-73b2a8de20b5


## Scroll from design prototype


https://github.com/Automattic/studio/assets/779993/afbad601-b995-490d-ba00-506cf3589eb6


### Regressions comparison

Here is how I compared the versions and couldn't find any regression 🙌 .

https://github.com/Automattic/studio/assets/779993/ed41a0d8-fbdc-46be-8253-6602b341a9e3

dcalhoun commented on src/components/content-tab-assistant.tsx:
```diff
@@ -1,44 +1,115 @@
+export const UserMessage = ( { text }: { text: string } ) => (
+	<div className="flex justify-end mb-2 mt-2">
+		<div className="inline-block p-2 rounded-sm border border-gray-300 lg:max-w-[70%] user-select-text">
+			{ text }
```
To improve the comprehensibility of this interface for users relying upon assitive technologies, we might consider labeling each message with the author. E.g., `aria-label="Your message, Install a theme"`, `aria-lable="Studio AI, You can install..."`. 

Additionally, we might consider announcing new messages as they arrive using [`@wordpress/a11y`](https://github.com/WordPress/gutenberg/tree/18ca36c0102b1384f0392e8d9d9e3db9c0f7cc02/packages/a11y#accessibility-a11y), which we [already leverage](https://github.com/search?q=repo%3AAutomattic%2Fstudio%20%40wordpress%2Fa11y&type=code) in the Studio project.

derekblank replied:
Good point. I think we should thoroughly consider the entire a11y experience for the chat interface implementation. 

@sejas What do you think about creating a separate issue to check and implement accessibility standards (if one does not already exist)?

sejas replied:
Yes!, I've added @dcalhoun notes in a new issue: https://github.com/orgs/Automattic/projects/543/views/29?pane=issue&itemId=65205240

dcalhoun commented on src/hooks/use-assistant.ts:
```diff
@@ -0,0 +1,34 @@
+	const addMessage = ( content: string, role: 'user' | 'assistant' ) => {
+		setMessages( ( prevMessages ) => {
+			const updatedMessages = [ ...prevMessages, { content, role } ];
+			localStorage.setItem( selectedSiteId, JSON.stringify( updatedMessages ) );
```
> It [updater function] must be pure, should take the pending state as its only argument, and should return the next state.

My understanding is that React [updater functions](https://react.dev/reference/react/useState#setstate) "must" be pure. 

My perception is that this local storage mutation side effect in the state updater function is likley innocuous, but there's a possibility could cause cryptic issues in the future. 

Is there an alternative approach that would allow us to avoid this side effect within the updater function without adding too much complexity or indirection? 

I suppose we could sync local storage in an effect? 

These thoughts would apply to the `clearMessages` function as well.

derekblank replied:
Thanks for the feedback on this section. This initial hook was intended to be iterative, but I agree we should explore avoiding the side effects from the updater function and local storage syncing. To reduce development complexity, I think it may be worthwhile to wait until implementing some of the API calls (at least partially), or remove some of the simulated server responses that were included here to more closely match the ultimate intended experience.

dcalhoun reviewed (commented):
I did not notice any specific issues while testing. Well done! 

I left a few suggestsions for consideration. 

> [...] it would be great if we can make the messages to scroll instead of the whole window.

I agree with @sejas that scrolling only the chat contents within the tab was my expectation. I think it's likely worth pursuing in this PR or a future one.

derekblank committed:
Update chat interface input to use sticky bottom-0

derekblank committed:
Replace UserMessage and ResponseMessage with Message component

derekblank committed:
Remove overflow-hidden from <main /> element

derekblank committed:
Use Button component update input styling

derekblank committed:
Use Tailwind select-text for chat text selection

derekblank commented:
Thank you both for the review @sejas and @dcalhoun!

I've made some updates:

- Consolidated the inline CSS to prefer Tailwind classes when possible
- Replaced `<UserMessage />` and `<ResponseMessage />` with a single `<Message />` component
- Updated the Assistant tab to use the UI `<Button />` component
- Made some general minor style spacing updates
- Updated scroll behavior to that the header and tab navigation does not scroll with the chat window. (I was uncertain which behavior was preferred in https://github.com/Automattic/studio/pull/175#issuecomment-2134528046, so thank you for clarifying. 🙇 )

Let me know if you feel we should make any of the further updates mentioned in the PR comments, or move them to subsequent PRs.

derekblank committed:
Update Tab scroll behavior to allow tab panel content to scroll when overflow

dcalhoun reviewed (approved):
> Let me know if you feel we should make any of the further updates mentioned in the PR comments, or move them to subsequent PRs.

The latest changes look good to me. 

For future reference, here are a few additional a11y improvements we might make in later changes. 

<details><summary>A11y improvement opportunities</summary>

**Improve tab focus ring consistency:**
<img width="647" alt="Ideally tab focus ring always surrounds the tab text, currently the Assistant tab surrounds the icon only" src="https://github.com/Automattic/studio/assets/438664/1d96835d-6c0c-463f-bb96-7e41547d88f7">

**Hide decorative input icon from screen readers:**
<img width="643" alt="Currently, the assistant input includes an icon selectable by screen readers, its selectability is likely unnecessary and unhelpful, we might hide it from screen readers" src="https://github.com/Automattic/studio/assets/438664/bb4e61ac-d61d-41dd-8f28-754a51a6792e">

</details>

sejas reviewed (approved):
It looks great!

Thanks for the changes! ✨ 

- I created the issue to improve the A11y https://github.com/Automattic/dotcom-forge/issues/7520, it's not a priority right now, since we will be modifying the frontend in the next weeks.

I'll create new small issues to address specific changes, like the dropdown to clear the chat.


https://github.com/Automattic/studio/assets/779993/2970c355-e253-4858-a012-967eab6bd826

sejas committed:
Remove unused import

derekblank commented:
Thanks for the extra review. I've checked again in Windows and RTL layouts, and things are looking stable. Going to merge, and will address other issues in follow-up PRs.
